deploy/aws/java11Exec/src/main/java/poseidon/App.java

```diff
@@ -0,0 +1,136 @@
+    WebSocketMessageType type;
+    String data;
+    int exitCode;
+}
+
+/**
+ * Handler for requests to Lambda function.
+ * This Lambda function executes the passed command with the provided files in an isolated Java environment.
+ */
+public class App implements RequestHandler<APIGatewayV2WebSocketEvent, APIGatewayProxyResponseEvent> {
+
+    // gson helps parse the json objects.
+    private static final Gson gson = new Gson();
+
+    // gwClient is used to send messages back via the WebSocket connection.
+    private AmazonApiGatewayManagementApi gwClient;
+
+    // connectionID helps to identify the WebSocket connection that has called this function.
+    private String connectionID;
+
+    public static final String disableOutputHeaderKey = "disableOutput";
+
+    // disableOutput: If set to true, no output will be sent over the WebSocket connection.
+    private boolean disableOutput = false;
+
+    public APIGatewayProxyResponseEvent handleRequest(final APIGatewayV2WebSocketEvent input, final Context context) {
+        APIGatewayV2WebSocketEvent.RequestContext ctx = input.getRequestContext();
+        this.gwClient = AmazonApiGatewayManagementApiClientBuilder.standard()
+                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("https://" + ctx.getDomainName() + "/" + ctx.getStage(), Regions.EU_CENTRAL_1.getName()))
+                .build();
+        this.connectionID = ctx.getConnectionId();
+        this.disableOutput = input.getHeaders() != null && input.getHeaders().containsKey(disableOutputHeaderKey) && Boolean.parseBoolean(input.getHeaders().get(disableOutputHeaderKey));
+        AwsFunctionRequest execution = gson.fromJson(input.getBody(), AwsFunctionRequest.class);
+
+        try {
+            File workingDirectory = this.writeFS(execution.files);
+
+            ProcessBuilder pb = new ProcessBuilder(execution.cmd).redirectErrorStream(true);
+            pb.directory(workingDirectory);
+            Process p = pb.start();
+            InputStream stdout = p.getInputStream(), stderr = p.getErrorStream();
+            this.forwardOutput(p, stdout, stderr);
+            p.destroy();
+            return new APIGatewayProxyResponseEvent().withStatusCode(200);
+        } catch (Exception e) {
+            this.sendMessage(e.toString().getBytes(StandardCharsets.UTF_8), WebSocketMessageType.WebSocketOutputError, 0);
+            return new APIGatewayProxyResponseEvent().withBody(e.toString()).withStatusCode(500);
+        }
+    }
+
+    // writeFS writes the files to the local filesystem.
+    private File writeFS(Map<String, String> files) throws IOException {
+        File workspace = Files.createTempDirectory("workspace").toFile();
+        for (Map.Entry<String, String> entry : files.entrySet()) {
+            File f = new File(workspace, entry.getKey());
+
+            boolean ok = f.getParentFile().mkdirs();
+            if (!ok) throw new IOException("Cannot create parent directories.");
```

Comment:
Probably not an issue here, but `mkdirs()` will also return false if the directories already existed and thus were not generated. As we expect to have a clean environment, this case should rather happen. Still, it might be worth noticing.